### ISSUE TITLE
Ability to match on multiple/absence of class

### ISSUE THREAD

**Issue by titusfortner** [watir/watir#578]
This comes from the discussion on this issue:
https://github.com/cheezy/page-object/issues/429

Theoretically we could implement something like this with code like
```
//*[contains(concat(' ', @class, ' '), ' yes ') and not(contains(concat(' ', @class, ' '), ' no '))]
```

But this could cause issues with backward compatibility because right now Watir lets you select with more than one class (which Selenium's `:class` selector does not) based on XPath's string comparison.

Right now this:
```html
    <div class="a b c">Match me!</div>
```

gets matched by:
```ruby
browser.div(class: 'a')
browser.div(class: 'b')
browser.div(class: 'c')
browser.div(class: 'a b')
browser.div(class: 'b c')
browser.div(class: 'a b c')
```

but does not get matched by:
```ruby
browser.div(class: 'c b a')
browser.div(class: 'c b')
browser.div(class: 'c a')
browser.div(class: 'b a')
browser.div(class: 'a c')
```

I mostly think this is bad, but I also think it would break a lot of people's stuff if we changed it to raise exception if spaces are found. I guess this is the difference between treating "class" in html as separate objects vs an ordered string.

but what if we implemented a `:classes` locator (raise exception if both `class` & `classes` used at the same time)

```ruby
# 1
browser.element(classes: {true => ['a', 'b'], false => ['d']})
# 2
browser.element(classes: ['a', 'b', 'd' => false])
# 3
browser.element(classes: ['a', 'b', '!d'])
# 4
browser.element(classes: "a b !d")
 ```

Anyone have opinions? @chrismcmahon?

**Comment by chrismcmahon** [comment]
I like the :classes locator idea. I find myself working with elements sharing a mix of classes quite often.

**Comment by jkotests** [comment]
I was under the impression, in terms of applying CSS, that the order of class names in a class attribute value is irrelevant (https://stackoverflow.com/a/29695391/1200545). Personally, I would prefer to make the :class locator order insensitive. To me, all of the current not matching examples should match. It's always strange to have to explain that
`browser.div(class: 'a b')` will be different than `browser.div(class: 'b a')` when applying the CSS doesn't care.

While there could be backwards incompatibility, I would hope it does not affect many people. At least, I would hope that people are not relying on an order that a developer could change at any time. For the scenarios that are impacted, wouldn't they be able to switch to a regular expression and get the same behaviour? 

I would also suggest that we stick to just a :class locator rather than :class/:classes. It is easier to remember.

In terms of the absence of a class, if we are okay with the backwards incompatibility, would it be more concise if we prefix absent classes with an exclamation mark:

~~~~ruby
browser.element(class: 'a b !d')
~~~~

**Comment by titusfortner** [comment]
In terms of applying CSS, class order definitely doesn't matter. In terms of the XPath engine matching strings in the DOM, it does. :)

Heh, I deleted the paragraph where I discussed an implementation of `browser.element(class: 'a b !d')` because I hate breaking backward compatibility (even if it is actually a widely used bug), without being able to throw a warning or a deprecation notice. [The WebDriver W3C Spec](https://w3c.github.io/webdriver/webdriver-spec.html#locator-strategies) no longer includes `class` as an option, but the [JSON Wire Protocol](https://github.com/SeleniumHQ/selenium/wiki/JsonWireProtocol#post-sessionsessionidelement) specified "compound class names are not permitted." Selenium will continue to support that functionality with the same limitation.

As such, I think we should implement `:classes` and throw a deprecation notice for anyone using `:class` with spaces in the String. That would keep the behavior more consistent between Selenium & Watir (for whatever difference that actually makes), and be more clear and intentional about what is actually getting matched. I updated the original post with the 4 different options I see for syntax... I like the idea of the value being an Array instead of a String, especially since you can't just copy/paste the String from `class` to `classes` and expect to get the desired result.
```html
<div class="a b c">One</div>
<div class="b a c">Two</div>
```
```ruby
# Now
browser.div(class: "a c").text       # => "Two"
# Proposed
browser.div(classes: "a c").text     # => "One"
```

**Comment by titusfortner** [comment]
Merged with #579
